### ISSUE TITLE
[@storybook/angular] [Knobs] forceRerender doesn't trigger Constructor

### ISSUE THREAD

**Issue by Terrafire123** [storybookjs/storybook#9288]
There seems to be no ability to force a story to use Knobs to modify the constructor of its component. Various issues have been opened about this, but they all inevitably become stale, then closed without resolution.

**To Reproduce**
1. Create a component and story with the following code:

```
@Component({
  selector: 'mybutton',
  template:'<div>Sample</div>',
})
export class ButtonComponent implements OnInit {
  @Input() myText?: string;
  constructor() {
    console.log('I"m under construction!');
    console.log('Text is: ', this.myText);
  }
  ngOnInit() {
    console.log('I"m being initialized!');
    console.log('Text is: ', this.myText);
  }
}

```

```
import { FORCE_RE_RENDER } from '@storybook/core-events';
import { addons } from '@storybook/addons';
import { text, withKnobs, button } from '@storybook/addon-knobs';
import { ButtonComponent } from '../somewhere/component.ts';

export default { 
  title: 'Banner',
  decorators: [
    withKnobs
  ]
}
export const Sample = () => ({
  component: ButtonComponent,
  props: {
    myText: text('myText', 'Main Title'),
    button: button('Refresh', () => {
      console.log('Attempting to trigger refresh');
      addons.getChannel().emit(FORCE_RE_RENDER);
    })
  },
});
```

**Expected Result**
There should be a console.log with the current value of the Knob input for both the constructor and ngOnInit every time ForceRender is run.

Failing that, there should be **some** way to trigger the constructor if desired, using **the current value of the knobs**. 

**System:**
Environment Info:
  System:
    OS: Windows Server 2016 10.0.14393
    CPU: (4) x64 Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
  Binaries:
    Node: 10.16.0 - C:\Program Files\nodejs\node.EXE
    npm: 6.9.0 - C:\Program Files\nodejs\npm.CMD

  Package.json:
    "@storybook/addon-actions": "^5.2.8",
    "@storybook/addon-links": "^5.2.8",
    "@storybook/addon-notes": "^5.2.8",
    "@storybook/addons": "^5.2.8",
    "@storybook/angular": "^5.2.8",
    "@storybook/addon-knobs": "^5.2.8",
    "@storybook/cli": "^5.2.8",
Angular CLI: 8.3.20
Angular: 8.2.14
@angular-devkit/architect          0.803.20
@angular-devkit/build-angular      0.803.21
@angular-devkit/build-ng-packagr   0.803.21
@angular-devkit/build-optimizer    0.803.21
@angular-devkit/build-webpack      0.803.21
@angular-devkit/core               8.3.20
@angular-devkit/schematics         8.3.20
@angular/cdk                       8.2.3
@angular/cli                       8.3.20
@angular/flex-layout               8.0.0-beta.27
@angular/material                  8.2.3
@ngtools/webpack                   8.3.21
@schematics/angular                8.3.20
@schematics/update                 0.803.20
ng-packagr                         5.7.1
rxjs                               6.4.0
typescript                         3.5.3
webpack                            4.39.2

**Additional context**
It seems there has been some discussion about this in the past, but it inevitably ends in a "Stale" tag and the thread being locked.

https://github.com/storybookjs/storybook/issues/7002
https://github.com/storybookjs/storybook/issues/4760

Not a duplicate of:

https://github.com/storybookjs/storybook/issues/8825
https://github.com/storybookjs/storybook/issues/1736

Seems to be directly related to (Or even caused by):
https://github.com/storybookjs/storybook/pull/6094

**Comment by shilman** [comment]
@Terrafire123 the issues go stale because nobody is picking them up. Do you want to try to fix this issue?

**Comment by dfagan2** [comment]
Running into the same thing

**Comment by CodeByAlex** [comment]
@dfagan2 + @Terrafire123 Quick solution if you want a full refresh of the iframe:                 
refresh: button('Refresh', () => { document.location.reload(); })

It works like a charm!

**Comment by Terrafire123** [comment]
@CodeByAlex  @shilman 

Hey, CodeByAlex.

Thank you for the suggestion, but apparently the issue wasn't explained well, and it doesn't solve the problem with Knobs, **at all**. 

The expected result is that the page should be re-rendered using **the current value of the knobs**. 

The code you suggested reloads the page, but when it reloads the page, two problems occur.

1. Knobs reset to their default values!
2. Knobs still doesn't modify the inputs during the constructor or OnInit().

The code you provided unfortunately doesn't help, and doesn't appear to be any different than simply hitting the browser refresh button.

**Comment by CodeByAlex** [comment]
@Terrafire123 - I apologize, not sure what I was thinking you meant when I first posted this but your issue clearly defines the problem. I’m not sure of an answer for that at this time.

**Comment by jose-airspace** [comment]
@Terrafire123 @CodeByAlex this happens for me when using a react class component. When a knob changes the render function gets called but the constructor does not. 

Sample component:
```
class MyContainer extends React.Component {
  constructor(props) { // called only on first render
    super(props);
    this.store = props.store; // only happens with original value
  }

  render() { // called every time Device ID knob changed
    const { ping } = this.props.store; // works
    //const { ping } = this.store; // does not work (only has original value)

    return (
      <div>
        {ping.device_name}
      </div>
    );
  }
}
```

Sample story:

```
export const showingPing = () => {
  const device_name = text('Device ID', 'T84007');
  const ping = { device_name };
  const store = { ping };

  return <MyContainer store={store} /> // called every time Device ID knob changed
};
```

**Comment by tc4rdpp** [comment]
Did anyone get this forceReRender to work?

**Comment by shilman** [comment]
Hi gang, We’ve just released addon-controls in 6.0-beta!

Controls are portable, auto-generated knobs that are intended to replace addon-knobs long term.

- [Controls README](https://github.com/storybookjs/storybook/tree/next/addons/controls)
- [Storybook Controls Walkthrough](https://gist.github.com/shilman/69c1dd41a466bae137cc88bd2c6ef487)
- [Migration instructions](https://github.com/storybookjs/storybook/blob/next/addons/controls/README.md#how-do-i-migrate-from-addon-knobs)

Please upgrade and try them out today. Thanks for your help and support getting this stable for release!
